### PR TITLE
A variety of changes needed for correct operation

### PR DESCRIPTION
**Pull request by jchunkins**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Functional Changes

pkg/cli/mirror/fbc_operators.go

- remove findFBCConfig & replace with extractDeclarativeConfigFromImage This is necessary because the layer processing did not handle whiteout files. Switched to go-containerregistry which handles this correctly.
- modify getManifest so that it is fat manifest aware, and returns the first manifest in the list, or for single images, returns that manifest
- Removed functions that are no longer necessary: GetCatalogConfigPath, getConfigPathFromConfigLayer, UntarLayers

pkg/cli/mirror/create.go

- make createOlmArtifactsForOCI manifest list aware and discovers the first image in the list, or for single images, uses that image directly
- calls replacement function extractDeclarativeConfigFromImage
- stores full path to the artifact in a map so this path does not need to be recalculated later

pkg/cli/mirror/operator.go

- make use of operatorCatalogToFullArtifactPath map to obtain the full path to the olm_artifacts directory and its contents and remove getOperatorCatalogRef() since it is no longer necessary
- make writeLayout() use the image config file to discover the platform when adding images, otherwise fall back to linux/amd64

pkg/cli/mirror/options.go

- add operatorCatalogToFullArtifactPath map[string]string

pkg/image/image.go

- replace getManifest with getFirstDigestFromPath which now gets the first digest for manifest lists, or the image digest for single image
- change logging behavior in ParseReference() when path does not exist and invalidate the ID in that case.

Non-functional Changes

pkg/cli/mirror/mirror.go

- removed redundant AsRespoitory() call since RepositoryName() already does this
- use constant for olm_artifacts directory

pkg/cli/options.go

- use constant for workspace

all files
- add documentation and change formatting for long function signatures
- if any file used deprecated io/ioutil, replace with io or os package as necessary
- update unit tests as necessary (the bulk of the files are related to new test cases)

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

**Test Configuration**:
* Firmware version:
* Hardware:
* Toolchain:
* SDK:

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: pkg/cli/mirror/options.go [hunk 1]
```diff
@@ -14,35 +14,36 @@ import (
...
-	cancelCh         <-chan struct{}
-	once             sync.Once
-	continuedOnError bool
-	remoteRegFuncs   RemoteRegFuncs
+	cancelCh                          <-chan struct{}
+	once                              sync.Once
+	continuedOnError                  bool
+	remoteRegFuncs                    RemoteRegFuncs
+	operatorCatalogToFullArtifactPath map[string]string // stores temporary paths to declarative config directory key: OCI URI (e.g. oci://foo which originates with v1alpha2.Operator.Catalog) value: <current working directory>/olm_artifacts/<repo>/<config folder>
 }
```
Comment (lmzuccarelli) [inline pkg/cli/mirror/options.go:47]: +100 for adding useful comments on each flag

**Review by sherine-k — COMMENTED** [review]
Hi John,
Thank you for an overall great job explaining and commenting on the code to make this improvement easy. 
I have one question overall, and moving on to some testing. @zhouying7780  will also be helping us out

File: pkg/image/image.go [hunk 2]
```diff
@@ -103,45 +95,97 @@ func ParseReference(ref string) (TypedImageReference, error) {
...
+		if err != nil {
+			return nil, err
+		}
+		return &hash, nil
+	}
+
+	// if manifest has more than one entry, indicate that something is probably not right with this OCI layout, but don't error out
+	if len(idxManifest.Manifests) > 1 {
+		klog.Infof("more than one image reference found in OCI layout %s using first entry only", ref)
+	}
```
[the remark this answers is not in the document]
Comment (jchunkins) [reply pkg/image/image.go]: fixed in https://github.com/openshift/oc-mirror/pull/611/commits/ff1e4ff2e60edf994f576557a731694e4b7e6f3c

File: pkg/image/image.go [hunk 3]
```diff
@@ -103,45 +95,97 @@ func ParseReference(ref string) (TypedImageReference, error) {
...
+			// if its an image, get the config hash and return that value
+			img, err := ii.Image(descriptor.Digest)
+			if err != nil {
+				return nil, err
+			}
+			hash, err := img.ConfigName()
+			if err != nil {
+				return nil, err
+			}
+			return &hash, nil
```
[the remark this answers is not in the document]
Comment (jchunkins) [reply pkg/image/image.go]: fixed in https://github.com/openshift/oc-mirror/pull/611/commits/ff1e4ff2e60edf994f576557a731694e4b7e6f3c

File: pkg/image/builder/image_builder.go [hunk 4]
```diff
@@ -80,85 +81,196 @@ func (b *ImageBuilder) Run(ctx context.Context, targetRef string, layoutPath lay
 	if err != nil {
 		return err
 	}
-	idxManifest, err := idx.IndexManifest()
+	// make a copy of the original manifest for later
+	originalIdxManifest, err := idx.IndexManifest()
 	if err != nil {
 		return err
 	}
+	originalIdxManifest = originalIdxManifest.DeepCopy()
```
Comment (sherine-k) [inline pkg/image/builder/image_builder.go:89]: Hi @jchunkins 
Just a nit pick, and for me to be sure there's nothing that we are overwriting by mistake: Is this the equivalent to: 
```suggestion
	originalIdxManifest, err := idx.IndexManifest().DeepCopy()
	if err != nil {
		return err
	}

```

File: pkg/image/builder/image_builder.go [hunk 5]
```diff
@@ -80,85 +81,196 @@ func (b *ImageBuilder) Run(ctx context.Context, targetRef string, layoutPath lay
...
+					return nil, err
+				}
+				resultIdx = processedIdx
+				// making an assumption here that at any given point in the parent/child
+				// hierarchy, there's only a single image index entry
+				return resultIdx, nil
+			case types.DockerManifestSchema2:
+				v2format = true
+			case types.OCIManifestSchema1:
+				v2format = false
```
Comment (sherine-k) [inline pkg/image/builder/image_builder.go:127]: Hi @jchunkins 
Question about this: we set v2format to false , because we consider that any other media types are not container images and can be ignored. Is this correct? 
ie. v2format=false can only mean this is a oci manifest

File: pkg/cli/mirror/catalog_images.go [hunk 6]
```diff
@@ -85,6 +109,14 @@ func (o *MirrorOptions) rebuildCatalogs(ctx context.Context, dstDir string) (ima
 			ctlgRef := image.TypedImage{}
 			ctlgRef.Type = imagesource.DestinationRegistry
 			sourceRef, err := image.ParseReference(img)
+			// since we can't really tell if the "img" reference originated from an actual docker
+			// reference or from an OCI file path that approximates a docker reference, ParseReference
+			// might not lowercase the name and namespace values which is required by the
+			// docker reference spec (see https://github.com/distribution/distribution/blob/main/reference/reference.go).
+			// Therefore we lower case name and namespace here to make sure it's done.
+			sourceRef.Ref.Name = strings.ToLower(sourceRef.Ref.Name)
```
Comment (sherine-k) [inline pkg/cli/mirror/catalog_images.go:117]: If things are done right, sourceRef.OCIFBCPath should not be empty  and this can be a good indication that the image is OCI local

File: pkg/cli/mirror/fbc_operators.go [hunk 7]
```diff
@@ -537,89 +546,52 @@ func findFirstAvailableMirror(ctx context.Context, mirrors []sysregistriesv2.End
 	return "", finalError
 }
 
-// getManifest reads the manifest of the OCI FBC image
+// getManifest reads the manifest of the OCI FBC image. If imgSrc is a manifest
+// list, it will return the first manifest available in the list
 // and returns it as a go structure of type manifest.Manifest
 func getManifest(ctx context.Context, imgSrc types.ImageSource) (manifest.Manifest, error) {
```
Comment (sherine-k) [inline pkg/cli/mirror/fbc_operators.go:552]: getManifest doesn't seem to be called anymore. Did I miss something?

File: pkg/image/builder/image_builder.go (hunk 4 above)
Comment (jchunkins) [reply pkg/image/builder/image_builder.go:89]: Yes, this would be equivalent (at least from a pseudo code standpoint). I can't do that directly due to the two valued return (i.e. the error value).

File: pkg/image/builder/image_builder.go (hunk 5 above)
Comment (jchunkins) [reply pkg/image/builder/image_builder.go:127]: TL;DR: v2format=false means OCI formatted media types

In **_theory_** a manifest list can contain images that are of different media types. For example, one could be a docker **_v1_** image (i.e. `application/vnd.docker.distribution.manifest.v1+json`) and another docker **_v2s2_** image (`application/vnd.docker.distribution.manifest.v2+json`), or an OCI image (i.e. `application/vnd.oci.image.manifest.v1+json`) , but in practice mixing of different types never happens. Also the `v1` image type is really rare and not really something we need to worry about, but we handle this case (see below).

The switch statement covers all of the cases we're going to care about:

- "manifest list" types
  - process recursively
- application/vnd.docker.distribution.manifest.v2+json
  - set v2format to true
- application/vnd.oci.image.manifest.v1+json
  - set v2format to false
- default
  - error out since we hit a mediatype we don't support

As you can see in the code we're using `v2format` to pick the right layer media type, and eventually we use it again to make sure the index has the right media type based on what the images are so they match.

File: pkg/cli/mirror/catalog_images.go (hunk 6 above)
Comment (jchunkins) [reply pkg/cli/mirror/catalog_images.go:117]: ### Clarification

Actually in this case, the sourceRef comes from the `img` variable, which is the "repoPath" which could be:

- A valid docker ref reconstituted from the file system path
  - example: `foo.com/foo/bar/<id>`
- Some path for a OCI layout
  - example: `some/path/provided/by/user/<id>`

Because `img` does not have `oci:` prefix, the `ParseReference` function is going to parse this "repoPath" value as a docker reference, and the `sourceRef.OCIFBCPath` is never going to be populated.

### Discussion for how `sourceRef.OCIFBCPath` could be properly populated
So even if you wanted to add the `oci:` prefix to `img`, how would you know when to do so? Looking at the examples above, there's only one way to distinguish the difference between the two and you'd have to process the assumed "domain" portion of the path (please refer to the BNF below for terminology). You'd have to start making assumptions that if the first component in the path contains `.` or `:` then it's a docker ref and otherwise it's an OCI path.  

```
domain                          := host [':' port-number]
host                            := domain-name | IPv4address | \[ IPv6address \]	; rfc3986 appendix-A
domain-name                     := domain-component ['.' domain-component]*
domain-component                := /([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])/
```

This gets tricky as the examples show:

Examples that are easy enough to distinguish as docker refs based on just the "domain":
`foo.com/foo/bar/<id>`
`1.2.3.4/foo/bar/<id>`
`::ffff:102:304/foo/bar/<id>`

Examples that have easy workarounds you can hard code to test for (e.g. localhost):
`localhost/foo/bar/<id>`

Examples that the user could do and you'd have no way to tell they are valid domains or not:
`my.example/foo/bar/<id>`
`my:example/foo/bar/<id>`

I'd worry about false positives for detecting a valid docker reference given the last two examples.

File: pkg/cli/mirror/fbc_operators_test.go [hunk 8]
```diff
@@ -326,19 +265,69 @@ func TestGetRelatedImages(t *testing.T) {
...
+						require.NoError(t, err)
+						handleIndex(t, innerIdx)
+					}
+				}
 			}
-			relatedImages, err := getRelatedImages(filepath.Join(tmpdir, "configs"), c.packages)
+			imageIndex, err := c.configsPath.ImageIndex()
+			require.NoError(t, err)
+			handleIndex(t, imageIndex)
+			relatedImages, err := getRelatedImages(actualDir, c.packages)
```
Comment (sherine-k) [inline pkg/cli/mirror/fbc_operators_test.go:330]: This is the part that PR #627 OCPBUGS-13116 is also changing.
getRelatedImages should be called on the declarativeConfig that is already filtered by a previous call to render. Therefore, there is no need to start back from the image.

File: pkg/cli/mirror/operator.go [hunk 9]
```diff
@@ -177,8 +240,8 @@ func (o *OperatorOptions) renderDCFull(ctx context.Context, reg *containerdregis
 	ctlgRef := ctlg.Catalog //applies for all docker-v2 remote catalogs
 	if ctlg.IsFBCOCI() {
 		// initialize path where we assume the catalog config dir is <current working directory>/olm_artifacts/<repo>/<config folder>
-		ctlgRef, err = o.getOperatorCatalogRef(ctx, ctlg.Catalog)
-		if err != nil {
+		var ok bool
+		if ctlgRef, ok = o.operatorCatalogToFullArtifactPath[ctlg.Catalog]; !ok {
```
Comment (sherine-k) [inline pkg/cli/mirror/operator.go:244]: At this point, err is not yet defined.
We should return a `fmt.Errorf(...)` in this case

File: pkg/cli/mirror/fbc_operators.go (hunk 7 above)
Comment (jchunkins) [reply pkg/cli/mirror/fbc_operators.go:552]: Since the purpose of `findFirstAvailableMirror` is to simply "poke" at the provided image reference to see if it exists (by attempting to get the image source and fetching its manifest and simply checking err == nil), it does not matter which manifest you get back from the `getManifest` function. In other words, this call:

```go
_, _, err = regFuncs.getManifest(ctx, nil, imgsrc)
```

takes an types.ImageSource and calls `types.ImageSource.GetManifest(ctx, nil)`. For a docker image source it's going to fetch the manifest for the tag/sha stored in the image source (which could refer to an image or index). For an oci image source assuming that the reference is `oci:<path>` (and not `oci:<path>:<image>`) and there's only one item in the index.json, you get that manifest (which could refer to an image or index). Bear in mind that the way that  `findFirstAvailableMirror` is called, we're only dealing with docker image references, so the comments about OCI are moot.

Bottom line is that I did know this `getManifest` function was no longer used, but I left this in place since it could be useful. If you'd like I can simply remove it and its associated test case since it's not used anywhere at this point.

File: pkg/cli/mirror/operator.go (hunk 9 above)
Comment (jchunkins) [reply pkg/cli/mirror/operator.go:244]: good catch! fixed in https://github.com/openshift/oc-mirror/pull/611/commits/f871dbec7e8c02f27c0bac470bb8c62d27c8faf0

File: pkg/cli/mirror/operator.go [hunk 10]
```diff
@@ -262,8 +330,8 @@ func (o *OperatorOptions) renderDCDiff(ctx context.Context, reg *containerdregis
 
 	ctlgRef := ctlg.Catalog //applies for all docker-v2 remote catalogs
 	if ctlg.IsFBCOCI() {
-		ctlgRef, err = o.getOperatorCatalogRef(ctx, ctlg.Catalog)
-		if err != nil {
+		var ok bool
+		if ctlgRef, ok = o.operatorCatalogToFullArtifactPath[ctlg.Catalog]; !ok {
```
[the remark this answers is not in the document]
Comment (jchunkins) [reply pkg/cli/mirror/operator.go:335]: good catch! fixed in https://github.com/openshift/oc-mirror/pull/611/commits/f871dbec7e8c02f27c0bac470bb8c62d27c8faf0

File: pkg/cli/mirror/fbc_operators_test.go (hunk 8 above)
Comment (jchunkins) [reply pkg/cli/mirror/fbc_operators_test.go]: thanks for the heads up. I've rebased my PR on those changes.

File: pkg/cli/mirror/operator.go [hunk 11]
```diff
@@ -424,23 +513,7 @@ func (o *OperatorOptions) plan(ctx context.Context, dc *declcfg.DeclarativeConfi
...
-		if err != nil {
-
-			return nil, fmt.Errorf("unable to retrieve configs layer for image %s:\n%v\nMake sure this catalog is in OCI format", operatorCatalog, err)
-		}
-		// initialize path starting with <current working directory>/olm_artifacts/<repo>
-		catalogContentsDir := filepath.Join(artifactsPath, repo)
-		// initialize path where we assume the catalog config dir is <current working directory>/olm_artifacts/<repo>/<config folder>
-		ctlgConfigDir := filepath.Join(catalogContentsDir, configsLabel)
-		relatedImages, err := getRelatedImages(ctlgConfigDir, ic.Packages)
+		relatedImages, err := getRelatedImages(indexDir, ic.Packages)
```
[the remark this answers is not in the document]
Comment (jchunkins) [reply pkg/cli/mirror/operator.go]: thanks for the heads up. I've rebased my PR on those changes.

File: pkg/cli/mirror/fbc_operators.go (hunk 7 above)
Comment (sherine-k) [reply pkg/cli/mirror/fbc_operators.go:470]: You're right for the `types.ImageSource.GetManifest(ctx, nil)`
For the local `getManifest`, removing would help eliminate confusion about things. WDYT?
We could still find it again in repo history.

File: pkg/cli/mirror/operator.go [hunk 12]
```diff
@@ -472,16 +563,40 @@ func (o *OperatorOptions) plan(ctx context.Context, dc *declcfg.DeclarativeConfi
...
-			return nil, fmt.Errorf("error copying oci fbc catalog to layout directory: %v", err)
+		// obtain the source directory for the OCI content
+		ociSourcePath := v1alpha2.TrimProtocol(ctlgRef.OCIFBCPath)
+
+		// Now copy the individual components of the source OCI source to its layout dir destination.
+		// This is done to ensure that files/folders that are not part of the OCI layout specification
+		// are not copied.
+
+		// copyOCILayoutFileOrFolder will copy a file or folder that belongs to a oci layout
+		copyOCILayoutFileOrFolder := func(sourcePath, destinationPath, fileOrDir string) error {
```
Comment (aguidirh) [inline pkg/cli/mirror/operator.go:585]: One of the goals of future refactoring is to reduce the quantity of anonymous functions since they make the code less readable and it is difficult to follow when debug is required. 

Is it possible to define this func outside of plan and call it as a defined func?

File: pkg/cli/mirror/create.go [hunk 13]
```diff
@@ -166,14 +193,72 @@ func (o *MirrorOptions) createOlmArtifactsForOCI(ctx context.Context, cfg v1alph
...
+		}
+		rootIndexManifest, err := rootIndex.IndexManifest()
+		if err != nil {
+			return err
+		}
+
+		// attempt to find the first image reference in the layout...
+		// for a manifest list only search one level deep.
+		var img v1.Image
+	loop:
```
Comment (aguidirh) [inline pkg/cli/mirror/create.go:214]: Labels are optional for break statements, is there any specific reasons of using label (loop) here? If there is not specific reason, it is possible simplying the code removing these labels.

Comment (aguidirh) [inline pkg/cli/mirror/create.go:214]: Labels are optional for break statements, is there any specific reasons of using label (loop) here? If there is not specific reason, it is possible simplying the code removing these labels.


```suggestion
	for _, descriptor := range rootIndexManifest.Manifests {

			if descriptor.MediaType.IsIndex() {
				// follow the descriptor using its digest to get the referenced index and its manifest
				childIndex, err := rootIndex.ImageIndex(descriptor.Digest)
				if err != nil {
					return err
				}
				childIndexManifest, err := childIndex.IndexManifest()
				if err != nil {
					return err
				}

				// at this point, find the first image and store it for later if possible
				for _, childDescriptor := range childIndexManifest.Manifests {
					if childDescriptor.MediaType.IsImage() {
						img, err = childIndex.Image(childDescriptor.Digest)
						if err != nil {
							return err
						}
						// no further processing necessary
						break
					}
				}

			} else if descriptor.MediaType.IsImage() {
				// this is a direct reference to an image, so just store it for later
				img, err = rootIndex.Image(descriptor.Digest)
				if err != nil {
					return err
				}
				// no further processing necessary
				break
			}
		}
```

File: pkg/cli/mirror/operator.go (hunk 12 above)
Comment (jchunkins) [reply pkg/cli/mirror/operator.go:585]: Moved to standalone function

File: pkg/cli/mirror/create.go (hunk 13 above)
Comment (jchunkins) [reply pkg/cli/mirror/create.go:214]: Simply removing the loop label is insufficient for what this code needs to do. We need to stop processing as soon as we find an appropriate image. If I don't use a label I need to put other control structures in place to break out of the inner loop and then break out of the outer loop, or refactor this into its own function. So the label does serve a purpose. Let me know if you'd prefer that I refactor this.

Comment (aguidirh) [reply pkg/cli/mirror/create.go:214]: Labels are optional for break statements, is there any specific reasons of using label (loop) here? If there is not specific reason, it is possible simplying the code removing these labels.


```suggestion
	for _, descriptor := range rootIndexManifest.Manifests {

			if descriptor.MediaType.IsIndex() {
				// follow the descriptor using its digest to get the referenced index and its manifest
				childIndex, err := rootIndex.ImageIndex(descriptor.Digest)
				if err != nil {
					return err
				}
				childIndexManifest, err := childIndex.IndexManifest()
				if err != nil {
					return err
				}

				// at this point, find the first image and store it for later if possible
				for _, childDescriptor := range childIndexManifest.Manifests {
					if childDescriptor.MediaType.IsImage() {
						img, err = childIndex.Image(childDescriptor.Digest)
						if err != nil {
							return err
						}
						// no further processing necessary
						break
					}
				}

			} else if descriptor.MediaType.IsImage() {
				// this is a direct reference to an image, so just store it for later
				img, err = rootIndex.Image(descriptor.Digest)
				if err != nil {
					return err
				}
				// no further processing necessary
				break
			}
		}
```

Comment (aguidirh) [reply pkg/cli/mirror/create.go:214]: Hi @jchunkins,

I did not see about the inner loop, now it makes senses for me to keep the label, just ignore my suggestion.

File: pkg/cli/mirror/operator.go (hunk 12 above)
Comment (aguidirh) [reply pkg/cli/mirror/operator.go]: Thanks @jchunkins

File: pkg/image/builder/image_builder.go [hunk 14]
```diff
@@ -80,85 +81,196 @@ func (b *ImageBuilder) Run(ctx context.Context, targetRef string, layoutPath lay
...
+	/*
+		processImageIndex is a recursive function that allows for traversal of the hierarchy of
+		parent/child indexes that can exist for a multi arch image. There's always
+		at least one index at the root since this is an OCI layout that we're dealing with.
+		In theory there can be "infinite levels" of "index indirection" for multi arch images, but typically
+		its only two levels deep (i.e. index.json itself which is level one, and the manifest list
+		defined in the blobs directory, which is level two).
+	*/
+	var processImageIndex func(idx v1.ImageIndex) (v1.ImageIndex, error)
+	processImageIndex = func(idx v1.ImageIndex) (v1.ImageIndex, error) {
```
Comment (aguidirh) [inline pkg/image/builder/image_builder.go:100]: same comment as before, is it possible to have it in a standalone function? Only because of readability purpose

Comment (jchunkins) [reply pkg/image/builder/image_builder.go:100]: This one gets a little more complicated because the `v2format` value is updated along the way as needed. I'll have to include that as a pointer to the bool to maintain the same behavior.

Comment (jchunkins) [reply pkg/image/builder/image_builder.go]: changed in commit https://github.com/openshift/oc-mirror/pull/611/commits/01c51f41b2b92ad160d3581da6f87cd930921050

File: pkg/cli/mirror/fbc_operators.go (hunk 7 above)
Comment (jchunkins) [reply pkg/cli/mirror/fbc_operators.go]: Agreed with removing the dead code. Changed in commit https://github.com/openshift/oc-mirror/pull/611/commits/e371440b76d901fc44bbf7b05929321024b28ea0

**Comment by zhouying7780** [comment]
Done pre-merge test , no regression test found

File: pkg/image/builder/image_builder.go [hunk 15]
```diff
@@ -80,30 +103,140 @@ func (b *ImageBuilder) Run(ctx context.Context, targetRef string, layoutPath lay
...
-	idxManifest, err := idx.IndexManifest()
+	// make a copy of the original manifest for later
+	originalIdxManifest, err := idx.IndexManifest()
+	if err != nil {
+		return err
+	}
+	originalIdxManifest = originalIdxManifest.DeepCopy()
+
+	// process the image index for updates to images discovered along the way
+	resultIdx, err := b.processImageIndex(ctx, idx, &v2format, update, targetRef, layers...)
```
Comment (soltysh) [inline pkg/image/builder/image_builder.go:114]: Nit: in the long run I'd suggest avoiding passing parameter which are modified by function, but rather explicitly return that value.

**Review by soltysh — APPROVED** [review]
No major objections from my end to proceed with this as is.

**Review by sherine-k — COMMENTED** [review]
/lgtm

**Comment by sherine-k** [comment]
/retitle CFE-7838: A variety of changes needed for correct operation with multi architecture catalogs
@jchunkins That bug was internal (pre-merge) so there's no need to add it. I added a comment in JIRA to easily follow it.

**Comment by sherine-k** [comment]
/retitle CFE-783: A variety of changes needed for correct operation with multi architecture catalogs

**Comment by stbenjam** [comment]
GitHub says this has conflicts: `This branch has conflicts that must be resolved`

Not sure why the bot didn't apply the label.

**Comment by ashishth09** [comment]
@stbenjam I have rebased and resolved the conflicts

**Review by sherine-k — CHANGES_REQUESTED** [review]
Hi @jchunkins @ashishth09 
I took a very quick look today. I didn't see one of the changes done by @jchunkins in a previous merged PR.
Would you mind checking if the rebase went fine?
Thank you so much

**Comment by ashishth09** [comment]
@sherine-k , I tried to recreate the conflict this PR had in my own forked repository as I had a backup of John's branch, the only conflict I saw while rebasing was https://github.com/ashishth09/oc-mirror/pull/1/conflicts. Invited you and John to take look. John & I will go through this to see if I missed anything. Thank you for raising the concern.

**Comment by jchunkins** [comment]
@sherine-k I performed another rebase from main to make sure everything is in sync, and then did a line by line comparison of the files in `jchunkins:CFE-760-manifest-list-minimal` against the backup of the branch before the rebase happened, and as far as I can see everything is accounted for. Specifically the `validateMapping` function in this PR is using `getImageList`. 

Please let me know if there was something else that does not seem correct.

**Comment by sherine-k** [comment]
/lgtm
Thanks @jchunkins !

**Comment by dgoodwin** [comment]
/payload 4.14 nightly blocking

In general blocking is probably the better signal for pre-merge testing. (expensive, but worth if it we're worried this might cause regressions)

Hoping to return to this in one of the next batches.

**Comment by sherine-k** [comment]
Hi @dgoodwin 
Those fails are almost positively not related to this PR, especially since the failing tests don't use oc-mirror at all. 
Should we rerun them? 
Should we re-ask for the merge after they pass?
Or is this all covered by the queue process in another way?
